File: apps/desktop/src/api/schemas/attentionNotification.ts
import { z } from "zod";

import type {
  AttentionNotification,
  AttentionNotificationEvent,
  AttentionNotificationEventParams,
  AttentionNotificationTarget,
  AttentionNotificationTaskDetailFocus,
} from "../attentionNotifications";

const id = z.string().min(1);
const ids = z.array(id);
const nonEmptyIDs = ids.min(1);
const notificationKind = z.enum(["question", "approval", "interrupted_current_node"]);

const identifierSchema = z.object({ kind: notificationKind, uuid: id }).strict();
const questionFocusSchema = z.object({ kind: z.literal("question"), ask_ids: nonEmptyIDs }).strict();
const approvalFocusSchema = z.object({ kind: z.literal("approval"), approval_id: id }).strict();
const interruptedFocusSchema = z.object({ kind: z.literal("interrupted_current_node") }).strict();
const focusSchema = z.discriminatedUnion("kind", [questionFocusSchema, approvalFocusSchema, interruptedFocusSchema]);

function focus(value: z.infer<typeof focusSchema>): AttentionNotificationTaskDetailFocus {
  switch (value.kind) {
    case "question":
      return { kind: value.kind, askIDs: value.ask_ids };
    case "approval":
      return { kind: value.kind, approvalID: value.approval_id };
    case "interrupted_current_node":
      return { kind: value.kind };
  }
}

const targetSchema = z.discriminatedUnion("kind", [
  z
    .object({
      kind: z.literal("workflow_task"),
      project_id: id.optional(),
      workflow_id: id.optional(),
      task_id: id,
      task_short_id: id.optional(),
      task_title: z.string().optional(),
      session_id: id.optional(),
      current_node_id: id.optional(),
      current_node_branch_key: id.optional(),
      focus: focusSchema,
    })
    .strict()
    .transform(
      (value): AttentionNotificationTarget => ({
        kind: value.kind,
        projectID: value.project_id,
        workflowID: value.workflow_id,
        taskID: value.task_id,
        taskShortID: value.task_short_id,
        taskTitle: value.task_title,
        sessionID: value.session_id,
        currentNodeID: value.current_node_id,
        currentNodeBranchKey: value.current_node_branch_key,
        focus: focus(value.focus),
      }),
    ),
  z
    .object({ kind: z.literal("session_prompt"), session_id: id })
    .strict()
    .transform((value): AttentionNotificationTarget => ({ kind: value.kind, sessionID: value.session_id })),
]);

const questionStateSchema = z
  .object({
    prepared_ask_ids: ids,
    materialized_ask_ids: ids,
    current_unresolved_ask_ids: ids,
    skipped_ask_ids: ids,
    preview: z.string().optional(),
    display_count: z.number().int().nonnegative(),
    materialized_count: z.number().int().nonnegative(),
  })
  .strict()
  .transform((value) => ({
    preparedAskIDs: value.prepared_ask_ids,
    materializedAskIDs: value.materialized_ask_ids,
    currentUnresolvedAskIDs: value.current_unresolved_ask_ids,
    skippedAskIDs: value.skipped_ask_ids,
    preview: value.preview,
    displayCount: value.display_count,
    materializedCount: value.materialized_count,
  }));

const notificationPayloadSchema = z
  .object({
    id: identifierSchema,
    kind: notificationKind,
    occurred_at: id,
    revision: z.number().int().positive(),
    question: questionStateSchema.nullish(),
    approval: z.object({ approval_id: id, message: z.string().optional() }).strict().nullish(),
    interrupted_current_node: z
      .object({ message: z.string().optional(), reason: z.string().optional(), detail_json: z.string().optional() })
      .strict()
      .nullish(),
    target: targetSchema,
  })
  .strict();

type NotificationPayload = z.output<typeof notificationPayloadSchema>;

const notificationSchema: z.ZodType<AttentionNotification> = notificationPayloadSchema
  .superRefine(validateNotificationCoherence)
  .transform((value) => ({
    id: value.id,
    kind: value.kind,
    occurredAt: value.occurred_at,
    revision: value.revision,
    question: value.question ?? null,
    approval: value.approval == null ? null : { approvalID: value.approval.approval_id, message: value.approval.message },
    interruptedCurrentNode:
      value.interrupted_current_node == null
        ? null
        : {
            message: value.interrupted_current_node.message,
            reason: value.interrupted_current_node.reason,
            detailJSON: value.interrupted_current_node.detail_json,
          },
    target: value.target,
  }));

export const attentionNotificationEventSchema: z.ZodType<AttentionNotificationEvent> = z.discriminatedUnion("type", [
  z
    .object({ type: z.literal("pending"), sequence: z.number().int().positive(), source: z.enum(["live", "snapshot"]), pending: notificationSchema })
    .strict(),
  z
    .object({ type: z.literal("resolved"), sequence: z.number().int().positive(), source: z.enum(["live", "snapshot"]), id: identifierSchema, kind: notificationKind, occurred_at: id })
    .strict()
    .transform((value) => ({ ...value, occurredAt: value.occurred_at })),
  z
    .object({ type: z.literal("snapshot_complete"), sequence: z.number().int().positive(), source: z.literal("snapshot"), session_id: id })
    .strict()
    .transform((value) => ({ ...value, sessionID: value.session_id })),
]);

export const attentionNotificationEventParamsSchema: z.ZodType<AttentionNotificationEventParams> = z
  .object({ event: attentionNotificationEventSchema })
  .strict();

function sameIDs(left: readonly string[], right: readonly string[]): boolean {
  return left.length === right.length && left.every((value, index) => value === right[index]);
}

function validateNotificationCoherence(value: NotificationPayload, context: z.RefinementCtx): void {
  if (value.id.kind !== value.kind) {
    context.addIssue({ code: "custom", message: "notification kind mismatch" });
    return;
  }
  switch (value.kind) {
    case "question":
      validateQuestionNotification(value, context);
      return;
    case "approval":
      validateApprovalNotification(value, context);
      return;
    case "interrupted_current_node":
      validateInterruptedCurrentNodeNotification(value, context);
  }
}

function validateQuestionNotification(value: NotificationPayload, context: z.RefinementCtx): void {
  if (value.question == null) {
    context.addIssue({ code: "custom", message: "question state required" });
  }
  if (value.approval != null || value.interrupted_current_node != null) {
    context.addIssue({ code: "custom", message: "question notification has unrelated payload" });
  }
  if (value.target.kind !== "workflow_task" || value.target.focus.kind !== "question") {
    context.addIssue({ code: "custom", message: "question notification target mismatch" });
    return;
  }
  if (value.question != null && !sameIDs(value.question.currentUnresolvedAskIDs, value.target.focus.askIDs)) {
    context.addIssue({ code: "custom", message: "question notification IDs do not match target focus" });
  }
}

function validateApprovalNotification(value: NotificationPayload, context: z.RefinementCtx): void {
  if (value.approval == null) {
    context.addIssue({ code: "custom", message: "approval state required" });
  }
  if (value.question != null || value.interrupted_current_node != null) {
    context.addIssue({ code: "custom", message: "approval notification has unrelated payload" });
  }
  if (value.target.kind !== "workflow_task" || value.target.focus.kind !== "approval") {
    context.addIssue({ code: "custom", message: "approval notification target mismatch" });
    return;
  }
  if (value.approval != null && value.approval.approval_id !== value.target.focus.approvalID) {
    context.addIssue({ code: "custom", message: "approval notification ID does not match target focus" });
  }
}

function validateInterruptedCurrentNodeNotification(
  value: NotificationPayload,
  context: z.RefinementCtx,
): void {
  if (value.interrupted_current_node == null) {
    context.addIssue({ code: "custom", message: "interrupted-current-node state required" });
  }
  if (value.question != null || value.approval != null) {
    context.addIssue({ code: "custom", message: "interrupted-current-node notification has unrelated payload" });
  }
  if (
    value.target.kind !== "workflow_task" ||
    value.target.focus.kind !== "interrupted_current_node" ||
    value.target.currentNodeID === undefined
  ) {
    context.addIssue({ code: "custom", message: "interrupted-current-node notification target mismatch" });
  }
}

File: shared/clientui/attention_notification.go
package clientui

import "time"

type AttentionNotificationEventType string

const (
	AttentionNotificationEventPending          AttentionNotificationEventType = "pending"
	AttentionNotificationEventResolved         AttentionNotificationEventType = "resolved"
	AttentionNotificationEventSnapshotComplete AttentionNotificationEventType = "snapshot_complete"
)

type AttentionNotificationSource string

const (
	AttentionNotificationSourceLive     AttentionNotificationSource = "live"
	AttentionNotificationSourceSnapshot AttentionNotificationSource = "snapshot"
)

type AttentionNotificationKind string

const (
	AttentionNotificationKindQuestion       AttentionNotificationKind = "question"
	AttentionNotificationKindApproval       AttentionNotificationKind = "approval"
	AttentionNotificationKindInterruptedRun AttentionNotificationKind = "interrupted_run"
)

type AttentionNotificationTargetKind string

const (
	AttentionNotificationTargetWorkflowTask  AttentionNotificationTargetKind = "workflow_task"
	AttentionNotificationTargetSessionPrompt AttentionNotificationTargetKind = "session_prompt"
)

type AttentionNotificationFocusKind string

const (
	AttentionNotificationFocusQuestion       AttentionNotificationFocusKind = "question"
	AttentionNotificationFocusApproval       AttentionNotificationFocusKind = "approval"
	AttentionNotificationFocusInterruptedRun AttentionNotificationFocusKind = "interrupted_run"
)

type AttentionNotificationEvent struct {
	Sequence   uint64                         `json:"sequence"`
	Source     AttentionNotificationSource    `json:"source"`
	Type       AttentionNotificationEventType `json:"type"`
	Pending    *AttentionNotification         `json:"pending,omitempty"`
	ID         *AttentionNotificationID       `json:"id,omitempty"`
	Kind       AttentionNotificationKind      `json:"kind,omitempty"`
	OccurredAt *time.Time                     `json:"occurred_at,omitempty"`
	SessionID  string                         `json:"session_id,omitempty"`
}

type AttentionNotificationID struct {
	Kind AttentionNotificationKind `json:"kind"`
	UUID string                    `json:"uuid"`
}

type AttentionNotification struct {
	ID             AttentionNotificationID                   `json:"id"`
	Kind           AttentionNotificationKind                 `json:"kind"`
	OccurredAt     time.Time                                 `json:"occurred_at"`
	Revision       uint64                                    `json:"revision"`
	Question       *AttentionNotificationQuestionState       `json:"question,omitempty"`
	Approval       *AttentionNotificationApprovalState       `json:"approval,omitempty"`
	InterruptedRun *AttentionNotificationInterruptedRunState `json:"interrupted_run,omitempty"`
	Target         AttentionNotificationTarget               `json:"target"`
}

type AttentionNotificationQuestionState struct {
	PreparedAskIDs          []string `json:"prepared_ask_ids"`
	MaterializedAskIDs      []string `json:"materialized_ask_ids"`
	CurrentUnresolvedAskIDs []string `json:"current_unresolved_ask_ids"`
	SkippedAskIDs           []string `json:"skipped_ask_ids"`
	Preview                 string   `json:"preview,omitempty"`
	DisplayCount            int      `json:"display_count"`
	MaterializedCount       int      `json:"materialized_count"`
}

type AttentionNotificationApprovalState struct {
	TaskTransitionID string `json:"task_transition_id"`
	Message          string `json:"message,omitempty"`
}

type AttentionNotificationInterruptedRunState struct {
	RunID      string `json:"run_id"`
	Message    string `json:"message,omitempty"`
	Reason     string `json:"reason,omitempty"`
	DetailJSON string `json:"detail_json,omitempty"`
}

type AttentionNotificationTarget struct {
	Kind        AttentionNotificationTargetKind       `json:"kind"`
	ProjectID   string                                `json:"project_id,omitempty"`
	WorkflowID  string                                `json:"workflow_id,omitempty"`
	TaskID      string                                `json:"task_id,omitempty"`
	TaskShortID string                                `json:"task_short_id,omitempty"`
	TaskTitle   string                                `json:"task_title,omitempty"`
	SessionID   string                                `json:"session_id,omitempty"`
	RunID       string                                `json:"run_id,omitempty"`
	Focus       *AttentionNotificationTaskDetailFocus `json:"focus,omitempty"`
}

type AttentionNotificationTaskDetailFocus struct {
	Kind             AttentionNotificationFocusKind `json:"kind"`
	AskIDs           []string                       `json:"ask_ids,omitempty"`
	TaskTransitionID string                         `json:"task_transition_id,omitempty"`
	RunID            string                         `json:"run_id,omitempty"`
}
